catalog/core/catalog-core-localstorageprovider/src/main/java/org/codice/ddf/catalog/content/impl/FileSystemStorageProvider.java

```diff
@@ -388,88 +401,137 @@ public void rollback(StorageRequest request) throws StorageException {
   }
 
   private ContentItem readContent(URI uri) throws StorageException {
-    Path file = getContentFilePath(uri);
+    Path path = getContentFilePath(uri);
 
-    if (file == null) {
+    if (path == null) {
       throw new StorageException(
           "Unable to find file for content ID: " + uri.getSchemeSpecificPart());
     }
 
-    String filename = file.getFileName().toString();
-    String extension = FilenameUtils.getExtension(filename);
-    URI reference = null;
-    // if the file is an external reference, remove the reference extension
-    // so we can get the real extension
-    if (REF_EXT.equals(extension)) {
-      extension = FilenameUtils.getExtension(FilenameUtils.removeExtension(filename));
-      try {
-        reference = new URI(new String(Files.readAllBytes(file), Charset.forName("UTF-8")));
-
-        if (reference.getScheme() == null) {
-          file = Paths.get(reference.toASCIIString());
-        } else if (reference.getScheme().equalsIgnoreCase("file")) {
-          file = Paths.get(reference);
-        } else {
-          file = null;
-        }
-      } catch (IOException | URISyntaxException e) {
-        throw new StorageException(e);
+    String filename = path.getFileName().toString();
+
+    // resolve external reference if necessary, determine the extension, and retrieve an
+    // InputStream to the content
+    InputStream contentInputStream;
+    String extension;
+
+    try {
+      if (REF_EXT.equals(FilenameUtils.getExtension(path.getFileName().toString()))) {
+        // remove the external reference extension so we can get the real extension
+        extension =
+            FilenameUtils.getExtension(
+                FilenameUtils.removeExtension(path.getFileName().toString()));
+        contentInputStream = getInputStreamFromReference(path);
+      } else {
+        extension = FilenameUtils.getExtension(path.getFileName().toString());
+        contentInputStream = getInputStreamFromResource(path);
```

Comment:
✏️ This and line 424 in the `if` block are identical. Can be pulled out to right below the `if/else` at the end of the `try` block.

Reply:
Line 424 is
```
        contentInputStream = getInputStreamFrom**Reference**(path);
```
whereas line 427 is
```
        contentInputStream = getInputStreamFrom**Resource**(path);
```

Reply:
Man. My eyes crossed on that. Thanks.